pkg/common/cns-lib/volume/manager.go

```diff
@@ -275,40 +371,84 @@ func (m *defaultManager) CreateVolume(ctx context.Context, spec *cnstypes.CnsVol
+				_ = m.operationStore.StoreRequestDetails(ctx, volumeOperationDetails)
+			}
+			return nil, errors.New(msg)
 		}
 		volumeOperationRes := taskResult.GetCnsVolumeOperationResult()
 		if volumeOperationRes.Fault != nil {
 			fault, ok := volumeOperationRes.Fault.Fault.(cnstypes.CnsAlreadyRegisteredFault)
 			if ok {
 				log.Infof("Volume is already registered with CNS. VolumeName: %q, volumeID: %q, opId: %q",
 					spec.Name, fault.VolumeId.Id, taskInfo.ActivationId)
+				if isIdempotencyHandlingEnabled {
+					volumeOperationDetails = createRequestDetails(volNameFromInputSpec,
+						fault.VolumeId.Id, "", 0,
+						volumeOperationDetails.OperationDetails.TaskInvocationTimestamp,
+						task.Reference().Value, taskInfo.ActivationId,
+						taskInvocationStatusSuccess, "")
+					err = m.operationStore.StoreRequestDetails(ctx, volumeOperationDetails)
+					if err != nil {
+						log.Warnf("failed to store CreateVolume details with error: %v", err)
+					}
+				}
 				return &CnsVolumeInfo{
 					DatastoreURL: "",
 					VolumeID:     fault.VolumeId,
 				}, nil
 			}
-			// Remove the taskInfo object associated with the volume name when the
-			// current task fails. This is needed to ensure the sub-sequent create
-			// volume call from the external provisioner invokes Create Volume.
-			taskDetailsInMap, ok := volumeTaskMap[volNameFromInputSpec]
-			if ok {
-				taskDetailsInMap.Lock()
-				log.Debugf("Deleted task for %s from volumeTaskMap because the task has failed", volNameFromInputSpec)
-				delete(volumeTaskMap, volNameFromInputSpec)
-				taskDetailsInMap.Unlock()
+
+			if isIdempotencyHandlingEnabled {
+				volumeOperationDetails = createRequestDetails(volNameFromInputSpec, "", "", 0,
+					volumeOperationDetails.OperationDetails.TaskInvocationTimestamp,
+					task.Reference().Value, taskInfo.ActivationId,
+					taskInvocationStatusError, volumeOperationRes.Fault.LocalizedMessage)
+				err = m.operationStore.StoreRequestDetails(ctx, volumeOperationDetails)
+				if err != nil {
+					log.Warnf("failed to store CreateVolume details with error: %v", err)
+				}
+			} else {
+				// Remove the taskInfo object associated with the volume name when the
+				// current task fails. This is needed to ensure the sub-sequent create
+				// volume call from the external provisioner invokes Create Volume.
+				taskDetailsInMap, ok := volumeTaskMap[volNameFromInputSpec]
+				if ok {
+					taskDetailsInMap.Lock()
+					log.Debugf("Deleted task for %s from volumeTaskMap because the task has failed", volNameFromInputSpec)
+					delete(volumeTaskMap, volNameFromInputSpec)
+					taskDetailsInMap.Unlock()
+				}
 			}
 			msg := fmt.Sprintf("failed to create cns volume %s. createSpec: %q, fault: %q, opId: %q",
 				volNameFromInputSpec, spew.Sdump(spec), spew.Sdump(volumeOperationRes.Fault),
 				taskInfo.ActivationId)
 			log.Error(msg)
 			return nil, errors.New(msg)
 		}
+		if isIdempotencyHandlingEnabled {
```

Comment:
We have this check 8 times in the create volume method and IMHO the code is not easy to read. I think it will be difficult to maintain the code. Can we have a parallel createVolumeWithIdempotency method with idempotency support? Maybe the createVolume can call one of the methods depending on whether the FSS is enabled or not.

Reply:
Refactored the code in the latest commits to reduce the number of fss checks